andi/andi.py

```diff
@@ -1,62 +1,327 @@
+    get_union_args,
+    is_union,
+    get_globalns,
+    get_unannotated_params,
+)
+from andi.errors import CyclicDependencyError, NonProvidableError
 
 
 def inspect(func: Callable) -> Dict[str, List[Optional[Type]]]:
     """
     For each argument of the ``func`` return a list of possible types.
+    Non annotated arguments are also returned with an empty list of possible
+    types.
     """
     globalns = get_globalns(func)
     annotations = get_type_hints(func, globalns)
+    for name in get_unannotated_params(func, annotations):
+        annotations[name] = None
     annotations.pop('return', None)
-    annotations.pop('cls', None)  # FIXME: pop first argument of methods
+    annotations.pop('self', None)  # FIXME: pop first argument of methods
+    annotations.pop('cls', None)
     res = {}
     for key, tp in annotations.items():
         if is_union(tp):
             res[key] = get_union_args(tp)
+        elif tp is None:
+            res[key] = []
         else:
             res[key] = [tp]
     return res
 
 
-def to_provide(
-        arguments_or_func: Union[
-            Callable,
-            Dict[str, List[Optional[Type]]]
-        ],
-        can_provide: Union[
-            Container[Type],
-            Callable[[Optional[Type]], bool]
-        ]
-        ) -> Dict[str, Optional[Type]]:
+ContainerOrCallableType = Union[
+    Container[Callable],
+    Callable[[Callable], bool]
+]
+
+
+class KwargsSpec(Dict[str, Callable]):
     """
-    Return a dictionary ``{argument_name: type}`` with types which should
-    be provided.
+    kwargs specification. Dict with the name of the argument
+    and the callable that is required to build an instance for such argument.
+    """
+
+    def kwargs(self, instances: Mapping[Callable, Any]) -> Dict[str, Any]:
+        """
+        Build the kwargs dict based on the spec using the prebuilt
+        instances provided in the input dictionary.
+
+        :param instances: A dict with the already prebuilt dependencies keyed
+                          by its builder
+        :return: a kwargs dict ready to be passed to a callable
+        """
+        return {name: instances[cls] for name, cls in self.items()}
 
-    ``arguments_or_func`` should be either a callable, or
-    a result of andi.inspect call for a callable.
 
-    ``can_provide`` can be either a function which receives a type and
-    returns True if argument of such type can be provided, or a container
-    (e.g. a set) with supported types.
+Step = Tuple[Callable, KwargsSpec]
+
+
+class Plan(List[Step]):
     """
-    if callable(arguments_or_func):
-        arguments = inspect(arguments_or_func)
-    else:
-        arguments = arguments_or_func
+    The resultant plan of executing the ``plan`` function.
+    It contains a sequence of steps that should be executed in order because
+    they are dependant among them.
+
+    The plan is itself a list of tuples of type
+    ``Tuple[Callable, Dict[str, Callable]]``
```

Comment:
What do you think about mentioning KwargsSpec here as well?

Reply:
Fixed here: https://github.com/scrapinghub/andi/pull/8/commits/3b184539c2746cd80b22147ad9e93a5b359343ba